File: packages/modules/lodashx/typescript/src/date/diff.ts
const ONE_SECOND = 1000;
const ONE_MINUTE = ONE_SECOND * 60;
const ONE_HOUR = ONE_MINUTE * 60;
const ONE_DAY = ONE_HOUR * 24;

export const diffTime = (a: Date, b: Date): number => {
  return Math.abs(a.getTime() - b.getTime());
};

export const diffDays = (a: Date, b: Date): number => {
  return Math.floor(diffTime(a, b) / ONE_DAY);
};

export const diffHours = (a: Date, b: Date): number => {
  return Math.floor(diffTime(a, b) / ONE_HOUR);
};

export const diffMinutes = (a: Date, b: Date): number => {
  return Math.floor(diffTime(a, b) / ONE_MINUTE);
};

export const diffSeconds = (a: Date, b: Date): number => {
  return Math.floor(diffTime(a, b) / ONE_SECOND);
};

export const diff = (a: Date, b: Date) => {
  return {
    days: () => diffDays(a, b),
    hours: () => diffHours(a, b),
    minutes: () => diffMinutes(a, b),
    seconds: () => diffSeconds(a, b),
  };
};

File: packages/modules/lodash/extensions/go/core/date/diff.go
package date

import "time"

const (
	oneSecond = 1000
	oneMinute = oneSecond * 60
	oneHour   = oneMinute * 60
	oneDay    = oneHour * 24
)

func diffMillis(a, b time.Time) int64 {
	ms := a.Sub(b).Milliseconds()
	if ms < 0 {
		return -ms
	}
	return ms
}

func DiffTime(a, b time.Time) int64 { return diffMillis(a, b) }

func DiffDays(a, b time.Time) int64  { return diffMillis(a, b) / oneDay }
func DiffHours(a, b time.Time) int64  { return diffMillis(a, b) / oneHour }
func DiffMinutes(a, b time.Time) int64 { return diffMillis(a, b) / oneMinute }
func DiffSeconds(a, b time.Time) int64 { return diffMillis(a, b) / oneSecond }

type DiffResult struct {
	a, b time.Time
}

func Diff(a, b time.Time) DiffResult { return DiffResult{a, b} }
func (d DiffResult) Days() int64    { return DiffDays(d.a, d.b) }
func (d DiffResult) Hours() int64   { return DiffHours(d.a, d.b) }
func (d DiffResult) Minutes() int64 { return DiffMinutes(d.a, d.b) }
func (d DiffResult) Seconds() int64 { return DiffSeconds(d.a, d.b) }
